### PR TITLE
HHH-18335 Improve partial order algorithm in AnnotationMetadataSourceProcessorImpl

### PR DESCRIPTION
**Pull request by NathanQingyangXu**
This PR aims to improve the partial order sorting implementation in AnnotationMetadataSourceProcessorImpl in terms of readability and complexity.

The trick is to maintain the partial order during the phase of collecting of all ancestor classes annotated with `@Entity` or `@MappedSuperclass`, so we don't need a separate complex ordering phase based on recursion.

https://hibernate.atlassian.net/browse/HHH-18335

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/annotations/AnnotationMetadataSourceProcessorImpl.java [hunk 1]
```diff
@@ -263,58 +266,34 @@ public void processEntityHierarchies(Set<String> processedEntityNames) {
...
-			if ( original.contains( clazz ) ) {
-				if ( !newList.contains( clazz ) ) {
-					newList.add( clazz );
-				}
-				copy.remove( clazz );
+			if ( orderedXClasses.size() > startIndex ) {
+				Collections.reverse( orderedXClasses.subList( startIndex, orderedXClasses.size() ) );
 			}
 		}
+		return orderedXClasses;
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/annotations/AnnotationMetadataSourceProcessorImpl.java:296]: the diff was screwed up by github, ending up with more difficult code review

File: hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/annotations/AnnotationMetadataSourceProcessorImpl.java [hunk 2]
```diff
@@ -263,23 +264,14 @@ public void processEntityHierarchies(Set<String> processedEntityNames) {
...
-	}
-
-	private void insertMappedSuperclasses(List<XClass> original, List<XClass> copy) {
+	/**
+	 * @return a partially ordered list so entry's ancestors always show up earlier
+	 */
+	private List<XClass> orderAndFillHierarchy(List<XClass> classes) {
 		final boolean debug = log.isDebugEnabled();
-		for ( XClass clazz : original ) {
+		LinkedHashSet<XClass> orderedClasses = CollectionHelper.linkedSetOfSize( classes.size() * 2 );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/annotations/AnnotationMetadataSourceProcessorImpl.java:272]:
```suggestion
		LinkedHashSet<XClass> orderedClasses = CollectionHelper.linkedSetOfSize( classes.size() * 2 );
		ArrayList<XClass> clazzHierarchy = new ArrayList<>();
```

File: hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/annotations/AnnotationMetadataSourceProcessorImpl.java [hunk 3]
```diff
@@ -289,32 +281,29 @@ private void insertMappedSuperclasses(List<XClass> original, List<XClass> copy)
 				}
 			}
 			else {
-				copy.add( clazz );
+				if ( orderedClasses.contains( clazz ) ) {
+					continue;
+				}
+
+				LinkedList<XClass> clazzHierarchy = new LinkedList<>();
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/annotations/AnnotationMetadataSourceProcessorImpl.java:288]:
```suggestion
				clazzHierarchy.clear();
```

File: hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/annotations/AnnotationMetadataSourceProcessorImpl.java [hunk 4]
```diff
@@ -289,32 +281,29 @@ private void insertMappedSuperclasses(List<XClass> original, List<XClass> copy)
...
-						&& !reflectionManager.equals( superClass, Object.class )
-						&& !copy.contains( superClass ) ) {
+						&& !reflectionManager.equals( superClass, Object.class ) ) {
 					if ( superClass.isAnnotationPresent( Entity.class )
 							|| superClass.isAnnotationPresent( MappedSuperclass.class ) ) {
-						copy.add( superClass );
+						if ( orderedClasses.contains( superClass ) ) {
+							break;
+						}
+						clazzHierarchy.addFirst( superClass ); // ensure ancestors shows up earlier
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/annotations/AnnotationMetadataSourceProcessorImpl.java:299]:
```suggestion
						clazzHierarchy.add( superClass );
```

File: hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/annotations/AnnotationMetadataSourceProcessorImpl.java [hunk 5]
```diff
@@ -289,32 +281,29 @@ private void insertMappedSuperclasses(List<XClass> original, List<XClass> copy)
...
 					}
 					superClass = superClass.getSuperclass();
 				}
+				orderedClasses.addAll( clazzHierarchy );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/annotations/AnnotationMetadataSourceProcessorImpl.java:303]:
```suggestion
				// ensure ancestors shows up earlier
				for ( int i = clazzHierarchy.size() - 1; i >= 0; i-- ) {
					orderedClasses.add( clazzHierarchy.get( i ) );
				}
```

**Review by beikov — CHANGES_REQUESTED** [review]
Looks ok, but I would like to ask you to avoid the use of `LinkedList` and the multiple instance creations.

File: hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/annotations/AnnotationMetadataSourceProcessorImpl.java (hunk 4 above)
Comment (NathanQingyangXu) [reply hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/annotations/AnnotationMetadataSourceProcessorImpl.java]: changed as suggested. thanks!

**Pull request merged**